internal/scheduler/unused/scheduler.go

```diff
@@ -0,0 +1,415 @@
+	Deleted bool
+	// Jobs waiting to be scheduled.
+	Pq JobPQ
+	// QueuePQ priority.
+	priority float64
+	// QueuePQ index. Maintained by the heap interface methods.
+	index int
+}
+
+func (pq QueuePQ) Len() int {
+	return len(pq)
+}
+
+func (pq QueuePQ) Less(i, j int) bool {
+	return pq[i].priority < pq[j].priority
+}
+
+func (pq QueuePQ) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *QueuePQ) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*Queue)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *QueuePQ) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// Update modifies the priority of an item in the queue.
+func (pq *QueuePQ) Update(item *Queue, priority float64) {
+	item.priority = priority
+	heap.Fix(pq, item.index)
+}
+
+// Priority queue of jobs. Each queue has such a queue associated with it.
+type JobPQ []*Job
+
+// Job submitted to a queue.
+type Job struct {
+	Id string
+	// See events.proto.
+	Fragile bool
+	// Resource claims for this job, e.g., "CPU: 2000, RAM: "
+	// CPU is measured in millicores, RAM in bytes, and accelerators in millis.
+	Claims map[string]int64
+	// JobPQ priority.
+	priority float64
+	// JobPQ index. Maintained by the heap interface methods.
+	index int
+}
+
+func (pq JobPQ) Len() int {
+	return len(pq)
+}
+
+func (pq JobPQ) Less(i, j int) bool {
+	return pq[i].priority < pq[j].priority
+}
+
+func (pq JobPQ) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *JobPQ) Push(x interface{}) {
+	n := len(*pq)
+	item := x.(*Job)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *JobPQ) Pop() interface{} {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	item.index = -1 // for safety
+	*pq = old[0 : n-1]
+	return item
+}
+
+// Update modifies the priority of an item in the queue.
+func (pq *JobPQ) Update(item *Job, priority float64) {
+	item.priority = priority
+	heap.Fix(pq, item.index)
+}
+
+// scheduleForever continually schedules jobs.
+func (s *Scheduler) scheduleForever() {
+	for {
+		err := s.scheduleOne()
+		if err != nil {
+			fmt.Printf("error: %s\n", err)
+		}
+		// TODO: Look at the error code to determine any changes that need to be made.
+	}
+}
+
+// scheduleOne schedules one job, or returns an error.
+func (s *Scheduler) scheduleOne() error {
+	queue, err := s.selectQueue()
+	if err != nil {
+		return err
+	}
+
+	job, err := queue.selectJob()
+	if err != nil {
+		return err
+	}
+
+	lease, err := s.suggestLease(queue, job)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("suggested lease: %s\n", lease)
+	return nil
+}
+
+// selectQueue returns the highest priority queue from the scheduler priority queue.
+// Does not remove the queue from the priority queue.
+func (s *Scheduler) selectQueue() (*Queue, error) {
```

Comment:
I'd be tempted to rename this `PeekQueue` and return nil rather than an error if nothing available. The caller can then check the result.